### ISSUE TITLE
Error in log when using switch

### ISSUE THREAD

**Issue by JangoBritt** [bendodson/homebridge-applescript-file-lightbulb#1]
When switching on, or increasing the brightness on this script, I get a fatal error in the homebridge log

`
16:56:43 - Unhandled Error Detected :  Error: spawn osascript ENOENT
    at exports._errnoException (util.js:870:11)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:178:32)
    at onErrorNT (internal/child_process.js:344:16)
    at nextTickCallbackWith2Args (node.js:442:9)
    at Immediate._tickCallback [as _onImmediate] (node.js:356:17)
    at processImmediate [as _immediateCallback] (timers.js:383:17)
`

My config.json is:

```
`   {
        "accessory": "ApplescriptFileLightbulb",
        "name": "Thomas Lullabies",
        "on": "/Users/Adam Davidson/.homebridge/applescript/ThomasMusicOn.scpt",
        "off": "/Users/Adam Davidson/.homebridge/applescript/ThomasMusicOff.scpt",
        "brightness": "/Users/Adam Davidson/.homebridge/applescript/ThomasMusicVolume.scpt"

    }
```

`
[edit] removed script contents because irrelevant and tl;dr

**Comment by bendodson** [comment]
I'm going to guess this is because the filename has a space in it as the error you are seeing is missing file. Could you try the following:
1. Try something like "/Users/Adam\ Davidson/.homebridge/applescript/ThomasMusicOn.scpt" to escape the space
2. If that doesn't work, move the files to another directory so the file path has no space in it (i.e. /applescript/ThomasMusicOn.scpt)

**Comment by JangoBritt** [comment]
Tried both of those.

\ isn't a valid escape character for JSON.

I copied the scripts into a new folder /applescript, but still get the same error.

as an aside, I tried configuring an invalid filename in the config.json, and the error was the same, meaning it's nothing to do with the file path.

**Comment by JangoBritt** [comment]
any thoughts?

To confirm my above statement, I tried it again with a brand new AppleScript, and get the same result.

**Comment by bendodson** [comment]
I've double checked at this end since updating to Sierra and newer versions of Homebridge and all still working as expected.

The only thing I can think to suggest is that I notice you are using *.scpt files. I'm guessing these are the binary versions being exported by the AppleScript editor? Try just saving a text file containing the AppleScript as a *.applescript file and seeing if that works.

**Comment by JangoBritt** [comment]
That sounded promising, but unfortunately, no change.

using the config:

`{
    "accessory": "ApplescriptFileLightbulb",
    "name": "Lullabies",
    "on": "/applescript/teston2.applescript",
    "off": "/applescript/testoff2.applescript",
    "brightness": "/applescript/testvolume2.applescript"
}`

I still get the same error message. I also am on Sierra.

The files are in plaintext.

Is there any chance I'm missing prerequisite files or nodes?

I'm getting the same error whether I do ON/OFF, or set the "brightness"

**Comment by JangoBritt** [comment]
so ENOENT means Error No Entry... and this error is "15:05:36 - Unhandled Error Detected :  Error: spawn osascript ENOENT"

Does this mean that it can't find osascript on my mac?

this is strange, because if I type "osascript /applescript/teston2.applescript" in my terminal window, it runs successfully.

Is the module trying to force the directory of OSASCRIPT, rather than looking at $PATH?

**Comment by bendodson** [comment]
Just realised that I didn't actually put instructions for installing the Applescript plugin which is the likely issue.

Can you run the following then try again?

`npm install -g applescript`

**Comment by bendodson** [comment]
The project uses the 'applescript' npm module so all the osascript stuff is handled in that. You could try updating the index.js file to use node-osascript instead to see it that solves the problem. Details on that module are at https://www.npmjs.com/package/node-osascript - you'd need to manually install the module (`npm install -g node-osascript`), update the index.js file to use 'node-osascript' instead of 'applescript', and alter the calls to applescript.execFile to use the syntax of node-osascript i.e.

`applescript.execFile(command, {brightness: level}, done);`

**Comment by bendodson** [comment]
You may also want to look at https://github.com/TooTallNate/node-applescript and see if you can run commands with it.

**Comment by JangoBritt** [comment]
Just to confirm, (as it looks like you might have deleted the comment) without changing your module, I should run `npm install -g applescript` ?

because I tried applescript, and homebridge-applescript, and neither are working (even after a reboot)

**Comment by bendodson** [comment]
Sorry, I deleted the comment as I'd forgotten that installing this plugin via npm would automatically install the applescript component it required. I'd try the steps I outlined above regarding switching the applescript dependency for node-osascript as the error messages you are seeing are likely something within the applescript codebase, not my extension which sits on top of it.

**Comment by JangoBritt** [comment]
haven't really got any further with this, I'm afraid... I installed node-applescript, but it's still not finding osascript, returning Error: spawn osascript ENOENT when I trigger the command on homebridge, even though I can manually run osascript from any path in terminal

**Comment by oglodyte** [comment]
Have similar problem, but only when launching homebridge with launchctl. 

events.js:160
      throw er; // Unhandled 'error' event
      ^

Error: spawn osascript ENOENT
    at exports._errnoException (util.js:1026:11)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:193:32)
    at onErrorNT (internal/child_process.js:359:16)
    at _combinedTickCallback (internal/process/next_tick.js:74:11)
    at process._tickCallback (internal/process/next_tick.js:98:9)

If running homebridge from terminal, everything works fine

**Comment by oglodyte** [comment]
Fixed my problem by adding /usr/bin to PATH in com.homebridge.server.plist

**Comment by Guoer-xz** [comment]
I have the same problem, no matter how the path is set……:(
Error: osascript: /xxxxx.applescript'': No such file or directory

**Comment by JangoBritt** [comment]
Try saving the file as plain text only, and not encoded

**Comment by Guoer-xz** [comment]
thanks very much,text only is right.Now the problem I have is that the first operation is accurate and effective, and then the button is no longer responding, do not touch the button generated by this plugin, homebridge works well

**Comment by Guoer-xz** [comment]
I am sorry, is not no response, but will wait for a long time to respond to the operation of this brightness settings again, there will be no response waiting…

**Comment by JangoBritt** [comment]
Does the script take a long time to run?

I've noticed that mine can take quite a while to run sometimes, and it doesn't update the switch status until it's completed

**Comment by Guoer-xz** [comment]
Thank you for your answer，Thank you for your answer, now I can use this plugin to work, although still wrong

  {
                    "accessory": "ApplescriptFileLightbulb",
                    "name": "iTunes Volume",
                    "on": "/volume100.applescript",
                    "off": "/volume0.applescript",
                    "brightness": "/volume.applescript"
                    }

Now each operation will be executed and then immediately back to the switch off, and in the homebridge prompt.

Error: Error: /volume100.applescript:22:28: execution error: 不能获得“item 1 of {}”。 (-1728)

**Comment by JangoBritt** [comment]
Looks like a problem with the script itself. 

Also, I've noticed in the HomeBridge logs, that when you change the brightness, it executes the ON command too. It might be worth removing the ON script from the config.json, because when you press ON in the Home app, it will set brightness to 100 anyway

**Comment by bendodson** [comment]
Closing this as it was covering a range of issues (some of which were fixed by a reinstallation) and because a recent fix for issue #4 solves the issues that have been raised about long hanging periods if an on and off script weren't provided.
